packages/backend/src/execution/middleware.ts

```diff
@@ -39,3 +49,39 @@ export function middleware<TQuery = unknown, TContext = unknown>({
         transformQuery,
     };
 }
+
+export const aclMiddleware = middleware<AnyQuery, AnyContext>({
+    predicate: ({ query, context }) => {
+        const missingPermissions: string[] = [];
+
+        mapRecursive(query, (node) => {
+            if (isQueryWithPermissions(node)) {
```

Comment:
I think it makes more sense to refactor this as follows:

```ts
predicate: () => true // runs on all queries
```

```ts
transformQuery: (query, context) => {
    throwIfPermissionsMissing(query, context.permissions??[])
    return query // doesn't modify the query
}
```
Notice how this makes it much easier to understand what the middleware does:

The key questions you want to ask yourself are, to what queries does this middleware apply to? all of them.

What happens when the middleware is applied to a query? it might throw an error, otherwise it won't modify the query.